// @rspress/shared@1.47.1/dist/node-utils.mjs
import * as __WEBPACK_EXTERNAL_MODULE_node_path_c5b9b54f__ from "node:path";
import * as __WEBPACK_EXTERNAL_MODULE_node_url_e96de089__ from "node:url";
import * as __WEBPACK_EXTERNAL_MODULE_gray_matter_90c6cd0b__ from "gray-matter";
import * as __WEBPACK_EXTERNAL_MODULE__rsbuild_core_1b356efc__ from "@rsbuild/core";
const extractTextAndId = (title)=>{
    if (!title) return [
        '',
        ''
    ];
    const customIdReg = /\\?{#.*}/;
    const text = title.replace(customIdReg, '').trimEnd();
    const customId = title.match(customIdReg)?.[0]?.slice(2, -1) || '';
    return [
        text,
        customId
    ];
};
function getIconUrlPath(icon) {
    if (!icon) return;
    icon = icon.toString();
    if (icon.startsWith('file://')) icon = (0, __WEBPACK_EXTERNAL_MODULE_node_url_e96de089__.fileURLToPath)(icon);
    if (!__WEBPACK_EXTERNAL_MODULE_node_path_c5b9b54f__["default"].isAbsolute(icon)) return icon;
    return `/${__WEBPACK_EXTERNAL_MODULE_node_path_c5b9b54f__["default"].basename(icon)}`;
}
function getNodeAttribute(node, attrName, attribute) {
    const found = node.attributes.find((attr)=>'name' in attr && attr.name === attrName);
    return attribute ? found : found?.value;
}
function loadFrontMatter(source, filepath, root, outputWarning = false) {
    try {
        const { content, data } = (0, __WEBPACK_EXTERNAL_MODULE_gray_matter_90c6cd0b__["default"])(source);
        const rawFrontMatter = source.slice(0, source.length - content.length);
        const emptyLinesSource = rawFrontMatter.length ? `${rawFrontMatter.replace(/[^\n]/g, '')}${content}` : content;
        return {
            content,
            frontmatter: data,
            emptyLinesSource
        };
    } catch (e) {
        if (outputWarning) __WEBPACK_EXTERNAL_MODULE__rsbuild_core_1b356efc__.logger.warn(`Parse frontmatter error in ${__WEBPACK_EXTERNAL_MODULE_node_path_c5b9b54f__["default"].relative(root, filepath)}: \n`, e);
    }
    return {
        content: '',
        frontmatter: {},
        emptyLinesSource: source
    };
}
const castArray = (value)=>Array.isArray(value) ? value : [
        value
    ];
const mergeDocConfig = async (...configs)=>{
    const { mergeWith } = await import("lodash-es");
    return mergeWith({}, ...configs, (target, source)=>{
        const pair = [
            target,
            source
        ];
        if (pair.some((item)=>void 0 === item)) return;
        if (pair.some((item)=>Array.isArray(item))) return [
            ...castArray(target),
            ...castArray(source)
        ];
    });
};
export { extractTextAndId, getIconUrlPath, getNodeAttribute, loadFrontMatter, mergeDocConfig };
