// @rspress/theme-default/dist/node/source-build-plugin.js
import * as __WEBPACK_EXTERNAL_MODULE_node_module_ab9f2194__ from "node:module";
import * as __WEBPACK_EXTERNAL_MODULE_node_path_c5b9b54f__ from "node:path";
import * as __WEBPACK_EXTERNAL_MODULE_node_url_e96de089__ from "node:url";
const tailwind_config = {
    content: [
        './src/**/*.{js,ts,jsx,tsx}'
    ],
    darkMode: 'class',
    theme: {
        backgroundColor: (ctx)=>({
                ...ctx.theme('colors'),
                white: 'var(--rp-c-bg)',
                soft: 'var(--rp-c-bg-soft)',
                mute: 'var(--rp-c-bg-mute)'
            }),
        extend: {
            fontSize: {
                '3xl': '2rem',
                '2xl': '1.625rem',
                xl: '1.375rem',
                lg: '1.25rem'
            },
            borderRadius: {
                '4xl': '2rem'
            },
            breakpoints: {
                xs: '640px',
                sm: '768px',
                md: '960px',
                lg: '1280px'
            },
            maxWidth: {
                60: '15rem'
            },
            maxHeight: {
                60: '15rem'
            },
            colors: {
                brand: {
                    DEFAULT: 'var(--rp-c-brand)',
                    light: 'var(--rp-c-brand-light)',
                    dark: 'var(--rp-c-brand-dark)',
                    lighter: 'var(--rp-c-brand-lighter)',
                    darker: 'var(--rp-c-brand-darker)'
                },
                text: {
                    1: 'var(--rp-c-text-1)',
                    2: 'var(--rp-c-text-2)',
                    3: 'var(--rp-c-text-3)',
                    4: 'var(--rp-c-text-4)'
                },
                divider: {
                    DEFAULT: 'var(--rp-c-divider)',
                    light: 'var(--rp-c-divider-light)',
                    dark: 'var(--rp-c-divider-dark)'
                },
                gray: {
                    light: {
                        1: 'var(--rp-c-gray-light-1)',
                        2: 'var(--rp-c-gray-light-2)',
                        3: 'var(--rp-c-gray-light-3)',
                        4: 'var(--rp-c-gray-light-4)',
                        5: 'var(--rp-c-gray-light-5)'
                    }
                },
                dark: {
                    light: {
                        1: 'var(--rp-c-dark-light-1)',
                        2: 'var(--rp-c-dark-light-2)',
                        3: 'var(--rp-c-dark-light-3)',
                        4: 'var(--rp-c-dark-light-4)',
                        5: 'var(--rp-c-dark-light-5)'
                    }
                }
            }
        }
    }
};
const source_build_plugin_require = (0, __WEBPACK_EXTERNAL_MODULE_node_module_ab9f2194__.createRequire)(import.meta.url);
const ROOT_DIR = (0, __WEBPACK_EXTERNAL_MODULE_node_url_e96de089__.fileURLToPath)(new URL('../..', import.meta.url).href);
function SourceBuildPlugin() {
    return {
        name: 'theme-default:source-build',
        builderConfig: {
            resolve: {
                alias: {
                    'rspress/theme': __WEBPACK_EXTERNAL_MODULE_node_path_c5b9b54f__["default"].resolve(ROOT_DIR, './src')
                }
            },
            tools: {
                postcss: (_, { addPlugins })=>{
                    try {
                        addPlugins(source_build_plugin_require('tailwindcss')({
                            config: {
                                ...tailwind_config,
                                content: tailwind_config.content.map((item)=>__WEBPACK_EXTERNAL_MODULE_node_path_c5b9b54f__["default"].resolve(ROOT_DIR, item))
                            }
                        }));
                    } catch (_e) {}
                }
            }
        }
    };
}
export { SourceBuildPlugin };
