// @rspress/core@1.47.1/dist/runtime/App.js
import * as __WEBPACK_EXTERNAL_MODULE_react_jsx_runtime_225474f2__ from "react/jsx-runtime";
import * as __WEBPACK_EXTERNAL_MODULE__rspress_runtime_0abd3046__ from "@rspress/runtime";
import * as __WEBPACK_EXTERNAL_MODULE__rspress_shared_baa012d0__ from "@rspress/shared";
import * as __WEBPACK_EXTERNAL_MODULE__theme_75e53063__ from "@theme";
import * as __WEBPACK_EXTERNAL_MODULE_react__ from "react";
import * as __WEBPACK_EXTERNAL_MODULE_react_helmet_async_048d06c6__ from "react-helmet-async";
import * as __WEBPACK_EXTERNAL_MODULE_virtual_global_components_43dbb356__ from "virtual-global-components";
import * as __WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__ from "virtual-site-data";
import "virtual-global-styles";
var App_QueryStatus = /*#__PURE__*/ function(QueryStatus) {
    QueryStatus["Show"] = "1";
    QueryStatus["Hide"] = "0";
    return QueryStatus;
}({});
async function initPageData(routePath) {
    var _siteData_multiVersion;
    const matchedRoute = (0, __WEBPACK_EXTERNAL_MODULE__rspress_runtime_0abd3046__.pathnameToRouteService)(routePath);
    if (matchedRoute) {
        var _mod_default___RSPRESS_PAGE_META;
        const mod = await matchedRoute.preload();
        const pagePath = (0, __WEBPACK_EXTERNAL_MODULE__rspress_shared_baa012d0__.cleanUrl)(matchedRoute.filePath);
        const extractPageInfo = __WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"].pages.find((page)=>{
            const normalize = (p)=>p.replace(/\/$/, '').toLowerCase();
            return (0, __WEBPACK_EXTERNAL_MODULE__rspress_runtime_0abd3046__.isEqualPath)(normalize(page.routePath), normalize(matchedRoute.path));
        });
        const encodedPagePath = encodeURIComponent(pagePath);
        const meta = (null == (_mod_default___RSPRESS_PAGE_META = mod.default.__RSPRESS_PAGE_META) ? void 0 : _mod_default___RSPRESS_PAGE_META[encodedPagePath]) || {};
        const { toc = [], title = '', frontmatter = {}, ...rest } = __WEBPACK_EXTERNAL_MODULE__rspress_shared_baa012d0__.MDX_OR_MD_REGEXP.test(matchedRoute.filePath) ? meta : mod;
        return {
            siteData: __WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"],
            page: {
                ...rest,
                pagePath,
                ...extractPageInfo,
                pageType: (null == frontmatter ? void 0 : frontmatter.pageType) || 'doc',
                title,
                frontmatter,
                toc
            }
        };
    }
    let lang = __WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"].lang || '';
    let version = (null == (_siteData_multiVersion = __WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"].multiVersion) ? void 0 : _siteData_multiVersion.default) || '';
    if (__WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"].lang && 'undefined' != typeof window) {
        const path = location.pathname.replace(__WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"].base, '').split('/').slice(0, 2);
        if (__WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"].locales.length) {
            const result = __WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"].locales.find(({ lang })=>path.includes(lang));
            if (result) lang = result.lang;
        }
        if (__WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"].multiVersion.versions) {
            const result = __WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"].multiVersion.versions.find((version)=>path.includes(version));
            if (result) version = result;
        }
    }
    return {
        siteData: __WEBPACK_EXTERNAL_MODULE_virtual_site_data_fa42d4c0__["default"],
        page: {
            pagePath: '',
            pageType: '404',
            routePath: '/404',
            lang,
            frontmatter: {},
            title: '404',
            toc: [],
            version,
            _filepath: '',
            _relativePath: ''
        }
    };
}
function App({ helmetContext }) {
    const { setData: setPageData, data } = (0, __WEBPACK_EXTERNAL_MODULE_react__.useContext)(__WEBPACK_EXTERNAL_MODULE__rspress_runtime_0abd3046__.DataContext);
    const frontmatter = data.page.frontmatter || {};
    const { pathname, search } = (0, __WEBPACK_EXTERNAL_MODULE__rspress_runtime_0abd3046__.useLocation)();
    const query = new URLSearchParams(search);
    const GLOBAL_COMPONENTS_KEY = 'globalUIComponents';
    const hideGlobalUIComponents = false === frontmatter[GLOBAL_COMPONENTS_KEY] || "0" === query.get(GLOBAL_COMPONENTS_KEY);
    (0, __WEBPACK_EXTERNAL_MODULE_react__.useLayoutEffect)(()=>{
        async function refetchData() {
            try {
                const pageData = await initPageData(pathname);
                null == setPageData || setPageData(pageData);
            } catch (e) {
                console.log(e);
            }
        }
        refetchData();
    }, [
        pathname,
        setPageData
    ]);
    return /*#__PURE__*/ (0, __WEBPACK_EXTERNAL_MODULE_react_jsx_runtime_225474f2__.jsxs)(__WEBPACK_EXTERNAL_MODULE_react_helmet_async_048d06c6__.HelmetProvider, {
        context: helmetContext,
        children: [
            /*#__PURE__*/ (0, __WEBPACK_EXTERNAL_MODULE_react_jsx_runtime_225474f2__.jsx)(__WEBPACK_EXTERNAL_MODULE__theme_75e53063__["default"].Layout, {}),
            !hideGlobalUIComponents && __WEBPACK_EXTERNAL_MODULE_virtual_global_components_43dbb356__["default"].map((componentInfo, index)=>{
                if (Array.isArray(componentInfo)) {
                    const [component, props] = componentInfo;
                    return /*#__PURE__*/ __WEBPACK_EXTERNAL_MODULE_react__["default"].createElement(component, {
                        key: index,
                        ...props
                    });
                }
                return /*#__PURE__*/ __WEBPACK_EXTERNAL_MODULE_react__["default"].createElement(componentInfo, {
                    key: index
                });
            })
        ]
    });
}
export { App, App_QueryStatus as QueryStatus, initPageData };
